Update v1.Taint parser to accept the form `key:effect` and `key=:effect-`

dlipovetsky opened the pull request:
/kind bug

**What this PR does / why we need it**:
Updates v1.Taint parser to accept the form `key:effect`, which is the [string representation](https://github.com/kubernetes/kubernetes/blob/33b20f2/pkg/apis/core/taint.go#L33) of a v1.Taint that has a nil Value (the empty string).

Also updates the parser to accept the form `key=:effect-`, since `key=:effect` is a valid string representation of a v1.Taint with a nil Value.

- [x] Update code and tests in `pkg/util/taints/`
- [x] Update [kubectl's copy of the code and tests](https://github.com/kubernetes/kubernetes/pull/69446) in `pkg/kubectl/cmd/taint/`
- [x] Update `kubectl taint` usage
 
**Which issue(s) this PR fixes**:
Fixes #73249

**Does this PR introduce a user-facing change?**:
```release-note
Support parsing more v1.Taint forms. `key:effect`, `key=:effect-` are now accepted.
```

/cc @liggitt

Commit:
Update v1.Taint parser to accept `key:effect`, `key=:effect-`, `key`, and `key-` forms

Also add missing tests for `key=:value` form.

dlipovetsky commented:
/sig scheduling

dlipovetsky commented:
/cc @kubernetes/sig-scheduling-bugs
/cc @k82cn
/cc @ravisantoshgudimetla

neolit123 reviewed (commented):
thanks @dlipovetsky 
/priority important-longterm

could you please add a user facing release note instead of NONE? e.g.
> `Support parsing v1.Taint forms such as xx, yy`

dlipovetsky commented:
Thanks @neolit123! Added release note.

dlipovetsky commented:
I added two new invalid spec tests and fixed a bug in the updated parser. (I also verified that the original parser passed these new tests)

dlipovetsky commented:
Once the changes to `pkg/util/taints/` are finalized, I'll make corresponding changes to `pkg/kubectl/cmd/taint/`.

Huang-Wei reviewed (commented):
LGTM for the pkg/util/taints/ part.

kevin-wangzefeng commented:
LGTM for `pkg/util/taints/`, I think you can go ahead for the rest parts.

dlipovetsky commented:
Thanks @kevin-wangzefeng and @Huang-Wei.

After I updated `parseTaints` and `parseTaint` in `pkg/kubectl/cmd/taint/util.go`, an [existing test](https://github.com/kubernetes/kubernetes/blob/f9e102cac7b08551d978764e3ac62e68fc6a2666/pkg/kubectl/cmd/taint/taint_test.go#L164) in `pkg/kubectl/cmd/taint/taint_test.go` failed:
```
--- FAIL: TestTaint (0.03s)
    --- FAIL: TestTaint/node_has_two_taints_with_the_same_key_but_different_effect,_remove_all_of_them_with_wildcard (0.00s)
        taint_test.go:313: Recovered: error: invalid taint spec: dedicated
            See 'taint -h' for help and examples
        taint_test.go:332: node has two taints with the same key but different effect, remove all of them with wildcard: node not tainted
```
I did not realize that `kubectl taint` accepts _a key with no value or effect_ to remove a taint; the example in the kubectl usage is: 
```
  # Remove from node 'foo' all the taints with key 'dedicated'
  kubectl taint nodes foo dedicated-
``` 
This use case was introduced in https://github.com/kubernetes/kubernetes/pull/30590. To be backward compatible, this PR should not break this use case.

There was no test in `pkg/util/taints` for removing a taint using the `key-` form. The parser prior to this PR accepts this form, but the parser in this PR rejects this form. My plan is to:
1. Add a test to `pkg/util/taints` for removing a taint using the `key-` form
2. Update the parser in this PR to accept the `key-` form

dlipovetsky commented:
The last commit passes the tests in `pkg/util/taints` and `pkg/kubectl/cmd/taint/`.

I have changed `parseTaint` to accept the `key` form, and ensured in `ParseTaints` that a taint with no effect cannot be added. The alternative is for `parseTaint` to continue rejecting this form, and handling the form as a special case in `ParseTaints`. I prefer the former, because the code already passes around taints without values or effects, which implies that the `key` form is valid, although it must not be used to create a taint.

kevin-wangzefeng commented:
I'm fine with it. Empty/null value and effect should not block parsing, though taints with only keys indicated should not be used for creating.

Commit:
Copy changes from `pkg/util/taints` to `pkg/kubectl/cmd/taint`

Commit:
Update kubectl taint usage

- Explain that the value is optional.
- Add example of adding a taint with no value to kubectl taint usage.

dlipovetsky commented:
Thanks for the feedback @kevin-wangzefeng.

I've copied the changes to `pkg/kubectl/cmd/taint` and updated the kubectl taint usage.

dlipovetsky commented:
/retest

jayunit100 commented:
@derekwaynecarr @aveshagarwal @ravisantoshgudimetla ~ can one of you help us push this taints patch through ? would like to possibly add some more related patches in the near future as well. (needs /approve) :)

Huang-Wei commented:
@dlipovetsky can you squash the PR to 3 commits? - each commit mapped to each task.

And only elders in pkg/OWNERS can approve changes in `pkg/util/taints/` :)

dlipovetsky commented:
@Huang-Wei Good idea, will do. Didn't want to squash too early.

Should I choose someone from pkg/OWNERS at random? :smile:

dlipovetsky commented:
/cc @kubernetes/api-approvers   (for changes in `pkg/util/taints`)
/cc @kubernetes/sig-cli-maintainers   (for changes in `pkg/kubectl/cmd/taint`)

bsalamat commented on pkg/kubectl/cmd/taint/utils.go:
```diff
@@ -56,52 +66,51 @@ func parseTaints(spec []string) ([]corev1.Taint, []corev1.Taint, error) {
 
-// parseTaint parses a taint from a string. Taint must be of the format '<key>=<value>:<effect>'.
+// parseTaint parses a taint from a string, whose form must be either
+// '<key>=<value>:<effect>', '<key>:<effect>', or '<key>'.
```
Why `<key>` is considered a valid taint here? It looks like it gets rejected later in `parseTaints` when adding a taint, but it is accepted when removing a taint. If this behavior is desired, it should be documented in the comments and the release note.

bsalamat replied:
I just read [your comment](https://github.com/kubernetes/kubernetes/pull/74159#issuecomment-466575717) above. It is fine, but please update comments and the release note to point it out.

dlipovetsky replied:
I updated the comment to `ParseTaints`, since that's the exported method.

I do want to point out that the exported behavior has not changed. Prior to this PR, the `ParseTaints` function parsed some forms (including  `<key>`) itself, but delegated parsing of other forms to `parseTaint`. This PR moves all parsing to `parseTaint`.

Given this, do you think updating the release note is necessary?

bsalamat reviewed (commented):
Looks good and thanks for cleaning up parts of the logic, @dlipovetsky!
Just a few minor comments.

bsalamat commented on pkg/util/taints/taints.go:
```diff
@@ -121,11 +139,21 @@ func ParseTaints(spec []string) ([]v1.Taint, []v1.Taint, error) {
 				return nil, nil, err
 			}
+			// validate that the taint has an effect, which is required to add the taint
+			if newTaint.Effect == "" {
```
Please use `len(newTaint.Effect) == 0` here as well.

dlipovetsky replied:
Thanks, fixed.

ravisantoshgudimetla reviewed (approved):
/lgtm

Thanks for the PR @dlipovetsky 

/hold

I will wait for @bsalamat and @liggitt comments but feel free to remove hold once it they're fine.

liggitt commented:
looks reasonable at a quick glance, but I'll defer to @bsalamat for review

bsalamat commented:
I reviewed and it looks fine. I have only a few minor comments. If @dlipovetsky addresses them sooner, we can get it merged.

dlipovetsky commented on pkg/kubectl/cmd/taint/taint.go:
```diff
@@ -80,7 +80,10 @@ var (
-		kubectl taint node -l myLabel=X  dedicated=foo:PreferNoSchedule`))
+		kubectl taint node -l myLabel=X  dedicated=foo:PreferNoSchedule
+
+		# Add a taint with the key 'bar' and no value
```
Thanks. Fixed.

bsalamat replied:
Sure. Please fix the example.

dlipovetsky replied:
Fixed!

dlipovetsky commented on pkg/kubectl/cmd/taint/utils.go:
```diff
@@ -40,11 +40,21 @@ func parseTaints(spec []string) ([]corev1.Taint, []corev1.Taint, error) {
 				return nil, nil, err
 			}
+			// validate that the taint has an effect, which is required to add the taint
+			if newTaint.Effect == "" {
```
Fixed.

bsalamat replied:
Did you push your changes?

dlipovetsky replied:
My apologies, I didn't! I have pushed them now.

dlipovetsky commented:
@bsalamat Thanks a lot for your review. My apologies for taking a while to address your comments. I address them using separate commits for the code in `pkg/util/taints` and `pkg/kubectl/cmd/taint`, so that I can easily fixup the original commits for those packages before merging.

dlipovetsky commented:
I also added a test to verify that the empty string is an invalid Taint spec.

dlipovetsky commented:
Test failed due to known flake, see #74931

/retest

bsalamat commented:
Thanks, @dlipovetsky! Looks good. Could you please squash commits?

/lgtm

bsalamat commented:
I updated the release note. Please check it.

dlipovetsky commented:
@bsalamat Squashed.

As for the release note, I'm not sure it needs to mention the `key-` form, since that has always been accepted--this PR just changes where its parsed internally (please see https://github.com/kubernetes/kubernetes/pull/74159#discussion_r264408700).

bsalamat commented:
@dlipovetsky Thanks again. I fixed the release notes.

/lgtm
/approve

bsalamat commented:
@liggitt Back to you for approval.

dlipovetsky commented:
/assign @liggitt 

(So that this PR shows up in your dashboard :smile:)

liggitt commented:
/approve

dlipovetsky commented:
@ravisantoshgudimetla Please cancel the hold. Thanks!
